### PR TITLE
[SPARK-8070] [SQL] [PYSPARK] avoid spark jobs in createDataFrame

### PR DESCRIPTION
**Pull request by davies**
Avoid the unnecessary jobs when infer schema from list.

cc @yhuai @mengxr

### CODE REVIEW

File: python/pyspark/sql/context.py [hunk 1]
```diff
@@ -315,28 +335,19 @@ def createDataFrame(self, data, schema=None, samplingRatio=None):
         else:
             rdd = data
 
-        if schema is None:
-            schema = self._inferSchema(rdd, samplingRatio)
+        if not isinstance(schema, StructType):
```
Comment (rxin) [inline python/pyspark/sql/context.py]: i think we should only do this if schema is None. 

and if schema is not None, and is not a StructType, we should throw a TypeError

Comment (rxin) [inline python/pyspark/sql/context.py]: Actually we are doing more than that. Let's change it so if it is the wrong type, we will throw typerror.

File: python/pyspark/sql/context.py [hunk 2]
```diff
@@ -189,7 +189,27 @@ def registerFunction(self, name, f, returnType=StringType()):
                                             self._sc._javaAccumulator,
                                             returnType.json())
 
+    def _inferSchemaLocally(self, data):
```
Comment (rxin) [inline python/pyspark/sql/context.py]: annotate what type data is in docstring?

**Comment by davies** [comment]
cc @rxin @JoshRosen

**Comment by davies** [comment]
@rxin ping

**Comment by rxin** [comment]
Merging this. Thanks!

**Pull request closed without merging**